core/src/main/java/com/linecorp/armeria/common/HttpHeaders.java

```diff
@@ -13,227 +13,157 @@
  * License for the specific language governing permissions and limitations
  * under the License.
  */
-
 package com.linecorp.armeria.common;
 
-import static java.util.Objects.requireNonNull;
-
-import java.util.Iterator;
-import java.util.Map.Entry;
-
-import javax.annotation.Nullable;
+import java.util.function.Consumer;
 
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 
-import io.netty.handler.codec.Headers;
-import io.netty.util.AsciiString;
-
 /**
- * HTTP/2 headers.
+ * Immutable HTTP/2 headers.
+ *
+ * @see HttpHeadersBuilder
+ * @see RequestHeaders
+ * @see ResponseHeaders
  */
 @JsonSerialize(using = HttpHeadersJsonSerializer.class)
 @JsonDeserialize(using = HttpHeadersJsonDeserializer.class)
-public interface HttpHeaders extends HttpObject, Headers<AsciiString, String, HttpHeaders> {
-
-    /**
-     * An immutable empty HTTP/2 headers.
-     */
-    HttpHeaders EMPTY_HEADERS = new DefaultHttpHeaders(false, 0).asImmutable();
+public interface HttpHeaders extends HttpObject, HttpHeaderGetters {
 
     /**
-     * Returns new empty HTTP headers.
+     * Creates a new empty builder.
      */
-    static HttpHeaders of() {
-        return new DefaultHttpHeaders();
+    static HttpHeadersBuilder builder() {
+        return new HttpHeadersBuilder();
     }
 
     /**
-     * Returns new HTTP request headers.
+     * An empty {@link HttpHeaders}.
+     *
+     * @deprecated Use {@link #of()}.
      */
-    static HttpHeaders of(HttpMethod method, String path) {
-        return new DefaultHttpHeaders().method(method).path(path);
-    }
+    @Deprecated
+    HttpHeaders EMPTY_HEADERS = of();
 
     /**
-     * Returns new HTTP response headers.
+     * Returns an empty singleton {@link HttpHeaders}.
      */
-    static HttpHeaders of(int statusCode) {
-        return of(HttpStatus.valueOf(statusCode));
+    static HttpHeaders of() {
+        return DefaultHttpHeaders.EMPTY;
     }
 
     /**
-     * Returns new HTTP response headers.
+     * Returns a new {@link HttpHeaders} with the specified header.
      */
-    static HttpHeaders of(HttpStatus status) {
-        return new DefaultHttpHeaders().status(status);
+    static HttpHeaders of(CharSequence name, CharSequence value) {
+        return new HttpHeadersBuilder().add(name, value).build();
     }
 
     /**
-     * Returns new HTTP headers with a single entry.
+     * Returns a new {@link HttpHeaders} with the specified header. The value is converted into
+     * a {@link String} via {@link HttpHeadersBase#addObject(CharSequence, Object)}.
      */
-    static HttpHeaders of(AsciiString name, String value) {
-        return new DefaultHttpHeaders().add(name, value);
+    static HttpHeaders of(CharSequence name, Object value) {
```

Comment:
Might be overthinking, but I always get nervous when the same method name has parameters that inherit from each other (I think many lint tools would complain about it too).

Reply:
Also personally, I think it's extremely rare to use non-string header values in most user code, so the helpers might not be used much to warrant the potential confusion.

Reply:
> Might be overthinking, but I always get nervous when the same method name has parameters that inherit from each other (I think many lint tools would complain about it too).

You mean it can be confusing to have `of()` methods in both `HttpHeaders` and its subtypes, or to have many parameters names `name` and `value`?

> Also personally, I think it's extremely rare to use non-string header values in most user code, so the helpers might not be used much to warrant the potential confusion.

I find our test cases use `Integer` or `MediaType` as a header value pretty often, so I think it's pretty useful?

Reply:
I was thinking about the fact that two methods with the same name take similar arguments 

`of(CharSequence, CharSequence)`
`of(CharSequence, Object)`

I personally almost always control click in intellij just to make sure the correct method was called or get nervous 

I think it relates to this point of effective java

http://thefinestartist.com/effective-java/41

Reply:
I see. At worst it will be some performance loss due to using `addObject()` instead of `add()` and the end result will be same. Are you still concerned? Do you have a suggestion for a better name?

Reply:
`ofObjects` would be clear and not ambiguous. I don't feel so strongly though if it seems obtuse

Reply:
Let me leave them as they are, then. It seems OK to me.